mongo_connector/dest_mapping.py

```diff
@@ -0,0 +1,186 @@
+import logging
+import threading
+import re
+from mongo_connector import errors
+
+
+LOG = logging.getLogger(__name__)
+
+"""New structure to handle both plain and wildcard mapping namespaces dynamically.
+"""
+
+class DestMapping():
+    def __init__(self, namespace_set=[], ex_namespace_set=[], inpt={}):
+        # a dict containing plain mappings
+        self.plain = {}
+        # a dict containing wildcard mappings
+        self.wildcard = {}
+        # a dict containing reverse plain mapping
+        # because the mappings are not duplicated, so the values should also be unique
+        self.reverse_plain = {}
+        # a dict containing plain db mappings, db -> an array of mapped db
+        self.plain_db = {}
+        
+        # the input namespace_set and ex_namespace_set could contain wildcard
+        self.namespace_set = namespace_set
+        self.ex_namespace_set = ex_namespace_set
+        
+        self.lock = threading.Lock()
+        
+        # initialize
+        for ns in namespace_set:
+            if ns not in inpt.keys():
+                inpt[ns] = ns
+
+        for k, v in inpt.items():
+            self.set(k,v)
+            
+    def match(self, src, dst_pattern):
+        """If source string src matches dst, return the matchobject"""
+        d_arr = re.split('(\*)',dst_pattern)
+        r_arr = []
+        for x in d_arr:
+            if x != "*":
+                r_arr.append(re.escape(x))
+            else:
+                r_arr.append("([^*]+)")
+                
+        reg_pattern = re.compile(''.join(r_arr))
+        m = reg_pattern.match(src)
+        return m
+    
+    def match_set(self, plain_src, dst_arr):
+        for x in dst_arr:
+            if plain_src == x:
+                return True
+            if "*" in x:
+                m = self.match(plain_src, x)
+                if m:
+                    return True
+        
+        return False
+    
+    def replace(self, src_match, map_pattern):
+        """Given the matchobject src_match, replace corresponding '*' in map_pattern with src_match."""
+        num_match = src_match.lastindex
+        if num_match != map_pattern.count("*"):
+            LOG.error("The number of wildcards in the mapping %s should be %d" %(map_pattern, num_match))
+            return None
+        pattern_arr = re.split('(\*)', map_pattern)
+        map_arr = []
+        map_index = 1
+        for x in pattern_arr:
+            if x != "*":
+                map_arr.append(x)
+            else:
+                map_arr.append(src_match.group(map_index))
+                map_index += 1
```

Comment:
I think this can be made easier, too. If there's only one wildcard in the map_pattern, then we can just substitute the group from src_match into where the "*" is in map_pattern like this:

    wildcard_matched = src_match.group(1)
    return map_pattern.replace("*", wildcard_matched)